### PR TITLE
Address bug #21448 "Correctly Size Bounding Boxes for Polylines"

### PR DESCRIPTION
**Pull request by rickdaniels**
Changed PloyLineEntity code to calculate the scale vector
(which PolyLineEntity uses as the bouding box) and the registration point
(the offset of the bounding box from the first point in the Polyline) based
on the points in the PolyLineEntity so the bounding box is just big enough
to contain all the points in the line.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1947/HighFidelity-Beta-PR11046-3189cb90e43f431bd99fb01503160c4ec9660156.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1947/HighFidelity-Beta-PR11046-3189cb90e43f431bd99fb01503160c4ec9660156.exe

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1954/HighFidelity-Beta-PR11046-0c03b4ec53bd799eed02681c3805d24702612bcf.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1954/HighFidelity-Beta-PR11046-0c03b4ec53bd799eed02681c3805d24702612bcf.dmg

**Comment by cozza13** [comment]
build this please

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 1]
```diff
@@ -141,23 +140,71 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
 
+void PolyLineEntityItem::calculateScaleAndRegistrationPoint() {
+	glm::vec3 high(0.0f, 0.0f, 0.0f);
+	glm::vec3 low(0.0f, 0.0f, 0.0f);
+	for (int i = 0; i < _points.size(); i++) {
+		glm::vec3 point = _points.at(i);
+
+		if (point.x > high.x){
+			high.x = point.x;
+		}
```
Comment (ZappoMan) [inline libraries/entities/src/PolyLineEntityItem.cpp]: coding standard: brace placement. Should be `} else {`

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 2]
```diff
@@ -141,23 +140,71 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
+		else if (point.x < low.x) {
+			low.x = point.x;
+		}
+
+		if (point.y > high.y){
+			high.y = point.y;
+		}
```
Comment (ZappoMan) [inline libraries/entities/src/PolyLineEntityItem.cpp]: coding standard: brace placement. Should be `} else if {`

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 3]
```diff
@@ -141,23 +140,71 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
+		else if (point.y < low.y) {
+			low.y = point.y;
+		}
+
+		if (point.z > high.z){
+			high.z = point.z;
+		}
```
Comment (ZappoMan) [inline libraries/entities/src/PolyLineEntityItem.cpp]: coding standard: brace placement. Should be `} else if {`

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 4]
```diff
@@ -141,23 +140,71 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
+		result.x = fabsf(high.x) + fabsf(low.x);
+		result.y = fabsf(high.y) + fabsf(low.y);
+		result.z = fabsf(high.z) + fabsf(low.z);
+		SpatiallyNestable::setScale(result);
+
+		glm::vec3 point = _points.at(0);
+		glm::vec3 startPointInScaleSpace = point - low;
+		glm::vec3 newRegistrationPoint = startPointInScaleSpace / result;
+		EntityItem::setRegistrationPoint(newRegistrationPoint);
+	}
```
Comment (ZappoMan) [inline libraries/entities/src/PolyLineEntityItem.cpp]: coding standard: brace placement. Should be `} else {`

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 5]
```diff
@@ -141,23 +140,71 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
+	else {
+		// if Polyline has only one or fewer points, use default dimension settings
+		SpatiallyNestable::setScale(glm::vec3(1.0f, 1.0f, 1.0f));
+		EntityItem::setRegistrationPoint(glm::vec3(0.0f, 0.0f, 0.0f));
+		return;
```
Comment (ZappoMan) [inline libraries/entities/src/PolyLineEntityItem.cpp]: why is there a return here? it doesn't really exit early...

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2040/HighFidelity-Beta-PR11046-0c03b4ec53bd799eed02681c3805d24702612bcf.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows2016/2040/HighFidelity-Beta-PR11046-0c03b4ec53bd799eed02681c3805d24702612bcf.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2048/HighFidelity-Beta-PR11046-3e2dbe58ec06a6add7b9e767e20ae3f81a4348c6.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows2016/2048/HighFidelity-Beta-PR11046-3e2dbe58ec06a6add7b9e767e20ae3f81a4348c6.exe

**Comment by rickdaniels** [comment]
Testing procedure:
- paint a PolyLineEntity using fingerPaint or open a sandbox that already has PolyLineEntities in it.
- select the PolyLineEntity from the entities list in the Create page
- observe that the bounding box for the PolyLineEntity tightly encompasses all of its points.

**Comment by Nex-Pro** [comment]
Checking this Win10 and Vive

**Comment by Nex-Pro** [comment]
Bounding boxes are now closer to the size of the polyline entity. 
However the polylines in each case exceed the bounding boxes. See image.

hifi-snap-by--on-2017-07-28_01-48-49

Is this due to the way the polylines are defined. @sethalves  @rickdaniels

**Comment by themelissabrown** [comment]
fail.  there may be problems with clipping and with the entity server telling the right interfaces about the entity if the bounding box isn't all the way around all the visible parts of the entity

**Comment by rickdaniels** [comment]
Huh... Can I get the Polyline assets you have in that test, or tell me how you made them? All the ones I made myself worked fine.

**Comment by rickdaniels** [comment]
Could you try testing in desktop view? For some reason I can't access the list or properties pages of the tablet in VR mode, only in tablet. Perhaps there's an issue with showing the bounding boxes in VR, (although I can't imagine why)

**Comment by themelissabrown** [comment]
@rickdaniels I think there's probably a difference in the desktop vs VR view. I'm hoping to have a fix for your tablet issues so you can debug.

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 6]
```diff
@@ -92,13 +92,12 @@ bool PolyLineEntityItem::appendPoint(const glm::vec3& point) {
...
-    glm::vec3 halfBox = getDimensions() * 0.5f;
-    if ((point.x < -halfBox.x || point.x > halfBox.x) || (point.y < -halfBox.y || point.y > halfBox.y) || (point.z < -halfBox.z || point.z > halfBox.z)) {
-        qCDebug(entities) << "Point is outside entity's bounding box";
-        return false;
-    }
+
     _points << point;
     _pointsChanged = true;
+
+	calculateScaleAndRegistrationPoint();
```
Comment (sethalves) [inline libraries/entities/src/PolyLineEntityItem.cpp]: use 4 spaces rather than a tab

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 7]
```diff
@@ -141,23 +140,66 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
-            if ((point.x < -halfBox.x || point.x > halfBox.x) ||
-                (point.y < -halfBox.y || point.y > halfBox.y) ||
-                (point.z < -halfBox.z || point.z > halfBox.z)) {
-                qCDebug(entities) << "Point is outside entity's bounding box";
-                return;
-            }
-        }
         _points = points;
+
+		calculateScaleAndRegistrationPoint();
```
Comment (sethalves) [inline libraries/entities/src/PolyLineEntityItem.cpp]: tabs

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 8]
```diff
@@ -141,23 +140,66 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
+		} else if (point.z < low.z) {
+			low.z = point.z;
+		}
+	}
+	const float EPSILON = 0.0001f;
+	if (_points.size() > 1) {
+		// if all the points in the Polyline are at the same place in space, use default dimension settings
+		if ((low - high).length() < EPSILON) {
+			SpatiallyNestable::setScale(glm::vec3(1.0f, 1.0f, 1.0f));
+			EntityItem::setRegistrationPoint(glm::vec3(0.0f, 0.0f, 0.0f));
```
Comment (sethalves) [inline libraries/entities/src/PolyLineEntityItem.cpp]: default registration point should probably be glm::vec3(0.5f)

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 9]
```diff
@@ -141,23 +140,66 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
+
+		calculateScaleAndRegistrationPoint();
+
         result = true;
     });
 
     return result;
 }
 
+void PolyLineEntityItem::calculateScaleAndRegistrationPoint() {
```
Comment (sethalves) [inline libraries/entities/src/PolyLineEntityItem.cpp]: tabs here also -- re-indent to be spaces

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 10]
```diff
@@ -141,23 +140,66 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
+		SpatiallyNestable::setScale(result);
+
+		glm::vec3 point = _points.at(0);
+		glm::vec3 startPointInScaleSpace = point - low;
+		glm::vec3 newRegistrationPoint = startPointInScaleSpace / result;
+		EntityItem::setRegistrationPoint(newRegistrationPoint);
+	} else {
+		// if Polyline has only one or fewer points, use default dimension settings
+		SpatiallyNestable::setScale(glm::vec3(1.0f, 1.0f, 1.0f));
+		EntityItem::setRegistrationPoint(glm::vec3(0.0f, 0.0f, 0.0f));
```
Comment (sethalves) [inline libraries/entities/src/PolyLineEntityItem.cpp]: use default of glm::vec3(0.5f)

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 11]
```diff
@@ -141,23 +140,66 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
+	if (_points.size() > 1) {
+		// if all the points in the Polyline are at the same place in space, use default dimension settings
+		if ((low - high).length() < EPSILON) {
+			SpatiallyNestable::setScale(glm::vec3(1.0f, 1.0f, 1.0f));
+			EntityItem::setRegistrationPoint(glm::vec3(0.0f, 0.0f, 0.0f));
+			return;
+		}
+
+		glm::vec3 result;
+		result.x = fabsf(high.x) + fabsf(low.x);
```
Comment (sethalves) [inline libraries/entities/src/PolyLineEntityItem.cpp]: The values in low and high might be negative.  If you end up with low.x = -1 and high.y = 1, this math is going to be wrong.  I think the fabsf should be removed.  Maybe it should be just `result.x = high.x - low.x` ?

File: libraries/entities/src/PolyLineEntityItem.cpp [hunk 12]
```diff
@@ -141,23 +140,66 @@ bool PolyLineEntityItem::setLinePoints(const QVector<glm::vec3>& points) {
...
     });
 
     return result;
 }
 
+void PolyLineEntityItem::calculateScaleAndRegistrationPoint() {
+	glm::vec3 high(0.0f, 0.0f, 0.0f);
+	glm::vec3 low(0.0f, 0.0f, 0.0f);
+	for (int i = 0; i < _points.size(); i++) {
+		glm::vec3 point = _points.at(i);
```
Comment (sethalves) [inline libraries/entities/src/PolyLineEntityItem.cpp]: I some of the body of this for-loop could be replaced by using an AACube.  See libraries/shared/src/AACube.h

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2358/HighFidelity-Beta-PR11046-87139f5d822e197097905502079d4c8734395a52.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2358/HighFidelity-Beta-PR11046-87139f5d822e197097905502079d4c8734395a52.exe

**Comment by YuppyPlays** [comment]
Testing complete:

paint a PolyLineEntity using fingerPaint or open a sandbox that already has PolyLineEntities in it. **Check**
select the PolyLineEntity from the entities list in the Create page **Check**
observe that the bounding box for the PolyLineEntity tightly encompasses all of its points. **Fail** (See image below)
image

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2482/HighFidelity-Beta-PR11046-8d248f5ccd2763859c8964847db37cdcbc567bf1.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2482/HighFidelity-Beta-PR11046-8d248f5ccd2763859c8964847db37cdcbc567bf1.exe

**Comment by Nex-Pro** [comment]
Looking at this.

**Comment by Nex-Pro** [comment]
Polylines do not appear to be exceeding the bounding box.

However they are not closely encompassed by the bounding box.  Drew polylines of various widths.

See images.

hifi-snap-by--on-2017-08-15_01-21-13
hifi-snap-by--on-2017-08-15_01-22-22

series of images of same polyline entity.
hifi-snap-by--on-2017-08-15_01-20-34
hifi-snap-by--on-2017-08-15_01-16-00
hifi-snap-by--on-2017-08-15_01-18-52
hifi-snap-by--on-2017-08-15_01-19-34
hifi-snap-by--on-2017-08-15_01-20-25
hifi-snap-by--on-2017-08-15_01-20-32

**Comment by Nex-Pro** [comment]
I think this needs to be looked at by design.

**Pull request merged**